### PR TITLE
Adds support for Vertex Array Objects

### PR DESCRIPTION
**Pull request by jwerle**
This pull request adds support for vertex array objects. VAOs are created per command if attributes are given and if the underlying WebGL implementation provides support for vertex array objects through ~a native implementation or~ an extension (OES). The pull request includes tests and updates to API.md. This should address #312.

cc @mikolalysenko @Erkaman @vipyne

---
This change is [Reviewable](https://reviewable.io/reviews/regl-project/regl/411)

### CODE REVIEW

**Comment by mikolalysenko** [comment]
*[lib/core.js, line 2876 at r3](https://reviewable.io:443/reviews/regl-project/regl/411#-Kjp2hVoqRxkPzMmc1BW:-Kjp2hVoqRxkPzMmc1BX:b-dxlmly) ([raw file](https://github.com/regl-project/regl/blob/e8a73fe4f06e86cb4c0010837f21f3842ad22e75/lib/core.js#L2876)):*
> ```JavaScript
> 
>     if (vaoState.hasSupport && hasAttributes) {
>       draw('if(', VAO, '){', VAO, '.unbind(); }')
> ```

Do we need to do this?  We should maybe save the current VAO state and defer unbinding only if needed.  Unbinding eagerly would cost an extra call per each VAO enabled draw call.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/regl-project/regl/411)*

**Comment by mikolalysenko** [comment]
*[lib/vao.js, line 9 at r3](https://reviewable.io:443/reviews/regl-project/regl/411#-Kjp3zhEZz2sBOYFhcua:-Kjp3zhEZz2sBOYFhcub:b37z71f) ([raw file](https://github.com/regl-project/regl/blob/e8a73fe4f06e86cb4c0010837f21f3842ad22e75/lib/vao.js#L9)):*
> ```JavaScript
>   // fetch shared extension once
>   if (extension === null) {
>     extension = gl.getExtension(EXTENSION_NAME)
> ```

We should reuse the existing extension module for this.  Also for testing purposes we should be able to disable the OES_vertex_array_object extension.  I do think that the default behavior though should be use VAOs when available.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/regl-project/regl/411)*

**Comment by mikolalysenko** [comment]
Overall I think this is a good start, but it needs a few improvements before it is ready to merge.

One thing that I do like is that it is very conservative, since it doesn't try to solve the problem of applying vertex array objects to dynamic attributes.  However, I think that eventually this is going to need to be implemented.  Also I am not sure the criteria here for applying a VAO is totally clear.  We need to check for attributes that are defined in scopes or partially dynamic as well before setting up the vertex array object.

**Comment by mikolalysenko** [comment]
*[lib/core.js, line 2866 at r3](https://reviewable.io:443/reviews/regl-project/regl/411#-Kjp5gOLf0WxG9fN36Ub:-Kjp5gOMdfSFivZJunY-:b-ekhfpl) ([raw file](https://github.com/regl-project/regl/blob/e8a73fe4f06e86cb4c0010837f21f3842ad22e75/lib/core.js#L2866)):*
> ```JavaScript
>     }
> 
>     injectExtensions(env, draw)
> ```

We are going to end up rebinding all the attributes for the VAO each time the command is executed, thus negating the benefits of using a VAO.  We should set all this up ahead of time at command construction.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/regl-project/regl/411)*

**Comment by mikolalysenko** [comment]
*[test/vao.js, line 1 at r3](https://reviewable.io:443/reviews/regl-project/regl/411#-Kjp6T-OL6aARykkoAiX:-Kjp6T-Pd7vObo2UA4q_:b-xfosvx) ([raw file](https://github.com/regl-project/regl/blob/e8a73fe4f06e86cb4c0010837f21f3842ad22e75/test/vao.js#L1)):*
> ```JavaScript
> var createContext = require('./util/create-context')
> ```

We should be careful here.  In the browser the WebGL context gets reused across all test cases to avoid crashing the browser.  In order to make this work we need to be careful that the patching code reverts all changes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/regl-project/regl/411)*

**Comment by mikolalysenko** [comment]
Another thing we need to test is that all the old vertex attribute code works with both VAOs enabled/disabled.  I'm not sure the best way to do this right now.

**Comment by mikolalysenko** [comment]
*[lib/vao.js, line 12 at r3](https://reviewable.io:443/reviews/regl-project/regl/411#-Kjp7IN08KlJhRSAi34j:-Kjp7IN12f_vBV8sZEJ0:bd5wy3b) ([raw file](https://github.com/regl-project/regl/blob/e8a73fe4f06e86cb4c0010837f21f3842ad22e75/lib/vao.js#L12)):*
> ```JavaScript
>   }
> 
>   var hasNativeSupport = typeof gl.createVertexArray === 'function'
> ```

We don't need to consider this case since regl is WebGL 1.0 exclusive.  The only way VAOs are available is via the extension.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/regl-project/regl/411)*

**Comment by mikolalysenko** [comment]
*[lib/vao.js, line 111 at r3](https://reviewable.io:443/reviews/regl-project/regl/411#-Kjp7_wtyyURU0-C9NsE:-Kjp7_wtyyURU0-C9NsF:bfxbio3) ([raw file](https://github.com/regl-project/regl/blob/e8a73fe4f06e86cb4c0010837f21f3842ad22e75/lib/vao.js#L111)):*
> ```JavaScript
>     }
> 
>     this.destroy = function () {
> ```

Need to make sure this gets called when `regl.destroy()` is called.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/regl-project/regl/411)*

**Comment by mikolalysenko** [comment]
This needs a bit more work, but it is promising.  Great initiative on this!

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/regl-project/regl/411#-:-Kjp7kpBefPZHp-J3TL7:b-acoxd)*

**Comment by mikolalysenko** [comment]
Reviewed 4 of 8 files at r1, 2 of 3 files at r2, 1 of 2 files at r3.
Review status: 7 of 8 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/regl-project/regl/411)*

**Comment by jwerle** [comment]
Awesome review! I'll have fixed up shortly!

**Comment by jwerle** [comment]
*[lib/vao.js, line 111 at r3](https://reviewable.io:443/reviews/regl-project/regl/411#-Kjp7_wtyyURU0-C9NsE:-KkHHL6RWDYLD7d6uwGv:b-hcj5ay) ([raw file](https://github.com/regl-project/regl/blob/e8a73fe4f06e86cb4c0010837f21f3842ad22e75/lib/vao.js#L111)):*
<details><summary><i>Previously, mikolalysenko (Mikola Lysenko) wrote…</i></summary><blockquote>

Need to make sure this gets called when `regl.destroy()` is called.
</blockquote></details>

I believe this is handled in the `clearVAOs()` function which is called in the `regl.destroy()` function

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/regl-project/regl/411)*

**Comment by jwerle** [comment]
*[lib/vao.js, line 9 at r3](https://reviewable.io:443/reviews/regl-project/regl/411#-Kjp3zhEZz2sBOYFhcua:-KkHNk8L2kA52VUJ9a8-:b-177pw2) ([raw file](https://github.com/regl-project/regl/blob/e8a73fe4f06e86cb4c0010837f21f3842ad22e75/lib/vao.js#L9)):*
<details><summary><i>Previously, mikolalysenko (Mikola Lysenko) wrote…</i></summary><blockquote>

We should reuse the existing extension module for this.  Also for testing purposes we should be able to disable the OES_vertex_array_object extension.  I do think that the default behavior though should be use VAOs when available.
</blockquote></details>

hmm so how should we go about making this possible? Do we want to enable the extension by providing it in the extensions array in the webgl config? How should we make this on by default and also configurable to shut off?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/regl-project/regl/411)*

**Comment by jwerle** [comment]
*[lib/vao.js, line 12 at r3](https://reviewable.io:443/reviews/regl-project/regl/411#-Kjp7IN08KlJhRSAi34j:-KkHOTc6BRqhBt5uCCOo:b-896fix) ([raw file](https://github.com/regl-project/regl/blob/e8a73fe4f06e86cb4c0010837f21f3842ad22e75/lib/vao.js#L12)):*
<details><summary><i>Previously, mikolalysenko (Mikola Lysenko) wrote…</i></summary><blockquote>

We don't need to consider this case since regl is WebGL 1.0 exclusive.  The only way VAOs are available is via the extension.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/regl-project/regl/411)*

**Comment by jwerle** [comment]
*[test/vao.js, line 1 at r3](https://reviewable.io:443/reviews/regl-project/regl/411#-Kjp6T-OL6aARykkoAiX:-KkHOWgrugLs6X94HHc6:b-896fix) ([raw file](https://github.com/regl-project/regl/blob/e8a73fe4f06e86cb4c0010837f21f3842ad22e75/test/vao.js#L1)):*
<details><summary><i>Previously, mikolalysenko (Mikola Lysenko) wrote…</i></summary><blockquote>

We should be careful here.  In the browser the WebGL context gets reused across all test cases to avoid crashing the browser.  In order to make this work we need to be careful that the patching code reverts all changes.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/regl-project/regl/411)*

**Comment by jwerle** [comment]
*[lib/core.js, line 2876 at r3](https://reviewable.io:443/reviews/regl-project/regl/411#-Kjp2hVoqRxkPzMmc1BW:-KkHQdcmQGn3Av6CA4RX:b-896fix) ([raw file](https://github.com/regl-project/regl/blob/e8a73fe4f06e86cb4c0010837f21f3842ad22e75/lib/core.js#L2876)):*
<details><summary><i>Previously, mikolalysenko (Mikola Lysenko) wrote…</i></summary><blockquote>

Do we need to do this?  We should maybe save the current VAO state and defer unbinding only if needed.  Unbinding eagerly would cost an extra call per each VAO enabled draw call.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/regl-project/regl/411)*

**Comment by jwerle** [comment]
*[lib/core.js, line 2866 at r3](https://reviewable.io:443/reviews/regl-project/regl/411#-Kjp5gOLf0WxG9fN36Ub:-KkHRcvrqTMT8_ZlK3No:b-896fix) ([raw file](https://github.com/regl-project/regl/blob/e8a73fe4f06e86cb4c0010837f21f3842ad22e75/lib/core.js#L2866)):*
<details><summary><i>Previously, mikolalysenko (Mikola Lysenko) wrote…</i></summary><blockquote>

We are going to end up rebinding all the attributes for the VAO each time the command is executed, thus negating the benefits of using a VAO.  We should set all this up ahead of time at command construction.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/regl-project/regl/411)*

**Comment by jwerle** [comment]
bump

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/regl-project/regl/411#-:-KlL0j2Ghk2k2pwZQf5I:bb7dx2z)*

**Comment by jwerle** [comment]
@mikolalysenko in a fork of this, I've rewritten the VAO stuff..

* A VAO is created and shared per instance of "core"  if extension opt-in and attributes found in `parseArguments()`
* VAO is only bound when attributes are given in `emitAttributes`
* VAO is binded when a draw call is made, and then unbind in `emitDraw`

This seems to work well... what do you think?

**Comment by jwerle** [comment]
@kevzettler reopening this PR

**Comment by kevzettler** [comment]
Thanks, @jwerle What would you say are the outstanding blocking issues or unknowns here? A little hard to follow the discussion but looks like you made the majority of suggested improvements.

**Comment by jwerle** [comment]
@kevzettler see my comment on Aug 17. I pushed those changes mentioned into this PR. Take a look

**Comment by kevzettler** [comment]
Thanks for the response @jwerle 
 
I was trying to ask for your opinion on what is needed to get this merged?

I'm looking at the branch now. I pushed some fixes for `standardjs` style compliance. Unfortunately, tests within `test/vao.js` are broken. I'm working on repairing. Other than those issues it looks like your comment on Aug 17 sounds like you've made all the suggested changes

**Comment by mikolalysenko** [comment]
Sorry to leave this sitting so long.  I'm working through the back log right now.

**Comment by JonDum** [comment]
VAOs are sorely missed from regl's api. Hope this can get merged in soon!
